File: backend/src/lib/coil-usage.ts
import type { CoilStatus } from "@prisma/client";
import { prisma } from "./prisma.js";

export const CRITICAL_COIL_FIELDS = [
  "grade",
  "coating",
  "size",
  "weight",
  "mtcNumber",
] as const;

export interface CoilUsageInfo {
  coilNumber: string;
  status: CoilStatus;
  slittingRecords: number;
  sunrackReceipts: number;
  productionBatches: number;
  dispatches: number;
  siteInstallations: number;
  complaints: number;
  documents: number;
  hasTraceabilityLinks: boolean;
  canEditCriticalFields: boolean;
  canDelete: boolean;
  canArchive: boolean;
}

export async function getCoilUsage(coilNumber: string): Promise<CoilUsageInfo | null> {
  const coil = await prisma.coil.findUnique({
    where: { coilNumber },
    include: {
      _count: { select: { slittingRecords: true, documents: true } },
      slittingRecords: {
        select: {
          sunrackReceipt: { select: { id: true } },
          batchConsumptions: {
            select: {
              batchNumber: true,
              batch: {
                select: {
                  dispatchLines: {
                    select: {
                      dispatchNoteNumber: true,
                      dispatch: {
                        select: { siteInstallation: { select: { id: true } } },
                      },
                    },
                  },
                  complaintLines: { select: { complaintId: true } },
                },
              },
            },
          },
        },
      },
    },
  });

  if (!coil) return null;

  let sunrackReceipts = 0;
  const batchSet = new Set<string>();
  const dispatchSet = new Set<string>();
  const installationSet = new Set<string>();
  const complaintSet = new Set<string>();

  for (const slit of coil.slittingRecords) {
    if (slit.sunrackReceipt) sunrackReceipts++;

    for (const consumption of slit.batchConsumptions) {
      batchSet.add(consumption.batchNumber);

      for (const line of consumption.batch.dispatchLines) {
        dispatchSet.add(line.dispatchNoteNumber);
        if (line.dispatch.siteInstallation) {
          installationSet.add(line.dispatch.siteInstallation.id);
        }
      }

      for (const complaintLine of consumption.batch.complaintLines) {
        complaintSet.add(complaintLine.complaintId);
      }
    }
  }

  const slittingRecords = coil._count.slittingRecords;
  const hasTraceabilityLinks = slittingRecords > 0;

  return {
    coilNumber: coil.coilNumber,
    status: coil.status,
    slittingRecords,
    sunrackReceipts,
    productionBatches: batchSet.size,
    dispatches: dispatchSet.size,
    siteInstallations: installationSet.size,
    complaints: complaintSet.size,
    documents: coil._count.documents,
    hasTraceabilityLinks,
    canEditCriticalFields: !hasTraceabilityLinks,
    canDelete: coil.status === "ACTIVE" && !hasTraceabilityLinks,
    canArchive: coil.status === "ACTIVE" && hasTraceabilityLinks,
  };
}

File: backend-go/internal/service/coil_usage.go
package service

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v5"

	"github.com/sunrack/ctrcms-go/internal/db"
)

var CriticalCoilFields = []string{"grade", "coating", "size", "weight", "mtcNumber"}

type CoilUsage struct {
	CoilNumber            string `json:"coilNumber"`
	Status                string `json:"status"`
	SlittingRecords       int64  `json:"slittingRecords"`
	SunrackReceipts       int64  `json:"sunrackReceipts"`
	ProductionBatches     int64  `json:"productionBatches"`
	Dispatches            int64  `json:"dispatches"`
	SiteInstallations     int64  `json:"siteInstallations"`
	Complaints            int64  `json:"complaints"`
	Documents             int64  `json:"documents"`
	HasTraceabilityLinks  bool   `json:"hasTraceabilityLinks"`
	CanEditCriticalFields bool   `json:"canEditCriticalFields"`
	CanDelete             bool   `json:"canDelete"`
	CanArchive            bool   `json:"canArchive"`
}

func GetCoilUsage(ctx context.Context, queries *db.Queries, coilNumber string) (*CoilUsage, error) {
	coil, err := queries.GetCoilByNumber(ctx, coilNumber)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, err
	}

	slitCount, err := queries.CountSlittingByCoil(ctx, coilNumber)
	if err != nil {
		return nil, err
	}
	docCount, err := queries.CountDocumentsByCoil(ctx, coilNumber)
	if err != nil {
		return nil, err
	}

	rows, err := queries.GetCoilUsageGraphRows(ctx, coilNumber)
	if err != nil {
		return nil, err
	}

	var sunrackSlits = make(map[string]struct{})
	batchSet := make(map[string]struct{})
	dispatchSet := make(map[string]struct{})
	installSet := make(map[string]struct{})
	complaintSet := make(map[string]struct{})

	for _, row := range rows {
		if row.SunrackReceiptID.Valid {
			sunrackSlits[row.SlitCoilId] = struct{}{}
		}
		if row.BatchNumber.Valid {
			batchSet[row.BatchNumber.String] = struct{}{}
		}
		if row.DispatchNoteNumber.Valid {
			dispatchSet[row.DispatchNoteNumber.String] = struct{}{}
		}
		if row.InstallationID.Valid {
			installSet[row.InstallationID.String] = struct{}{}
		}
		if row.ComplaintId.Valid {
			complaintSet[row.ComplaintId.String] = struct{}{}
		}
	}

	hasTrace := slitCount > 0

	return &CoilUsage{
		CoilNumber:            coil.CoilNumber,
		Status:                string(coil.Status),
		SlittingRecords:       slitCount,
		SunrackReceipts:       int64(len(sunrackSlits)),
		ProductionBatches:     int64(len(batchSet)),
		Dispatches:            int64(len(dispatchSet)),
		SiteInstallations:     int64(len(installSet)),
		Complaints:            int64(len(complaintSet)),
		Documents:             docCount,
		HasTraceabilityLinks:  hasTrace,
		CanEditCriticalFields: !hasTrace,
		CanDelete:             coil.Status == db.CoilStatusACTIVE && !hasTrace,
		CanArchive:            coil.Status == db.CoilStatusACTIVE && hasTrace,
	}, nil
}
